api/types/stacks/stacks.go

```diff
@@ -0,0 +1,91 @@
+package stacks // import "github.com/docker/docker/api/types/stacks
+
+import (
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/swarm"
+)
+
+// Stack is an object representing a server-side stack
+type Stack struct {
+	// ID is the ID of the stack, generated by the engine
+	ID string
+	swarm.Meta
+
+	// Spec is the StackSpec defining what this stack should look like
+	Spec StackSpec `json:",omitempty"`
+}
+
+// StackSpec defines the spec for a server-side stack.
+type StackSpec struct {
+	// StackSpecs include swarm-style annotations. The stack name cannot be
+	// changed once it has been created.
+	swarm.Annotations
+
+	// Services is a list of all service specs desired for this stack
+	Services []swarm.ServiceSpec `json:",omitempty"`
+
+	// Networks is a list of NetworkCreateRequest that define the networks
+	// required for this stack. NetworkCreateRequest includes the network name.
+	Networks []types.NetworkCreateRequest `json:",omitempty"`
+
+	// Secrets is a list of all Secrets required for this stack
+	Secrets []swarm.SecretSpec `json:",omitempty"`
+
+	// Configs is a list of all Configs required for this stack
+	Configs []swarm.ConfigSpec `json:",omitempty"`
+}
+
+// TODO(dperny): the following types should go in ../client.go, but are here
+// for ease of early review
+
+type StackCreateOptions struct {
+	// EncodedRegistryAuth is the encoded registry authorization credentials
+	// used for services in this stack. This is identical to the
+	// EncodedRegistryAuth field in ServiceCreateOptions and
+	// ServiceUpdateOptions
+	//
+	// This field follows the format of the X-Registry-Auth header.
+	EncodedRegistryAuth string
```

Comment:
Don't think this will work; if a stack contains multiple services, it's not unlikely some of those services use images from different registries (e.g. one from `docker.io/foo/bar` and one from `gcr.io/bar/baz`).

So either this should sent all credentials from the CLI (not very desirable), or perhaps we should think of https://github.com/moby/moby/issues/38591

Would these be stored as part of the service? (i.e., in case a task is re-deployed)

Reply:
I'm not sure if #38591 will help much here.
Probably the service spec will need something like an `ImagePullSecretRef` field and have that reference a secret... this of course will also be helpful for other scenarios (e.g. a login token times out and a task is rescheduled).

Reply:
How does the CLI currently decide which credentials to send? This problem is, I think, bigger than this particular feature, and is only coming to a head now that we have a situation in which multiple different registries may be used.

Reply:
The CLI creates each service and does `--with-registry-auth` to set the auth for each individual service when creating it. It just fetches the registered secret from the backing store.

Creating on the daemon side is a bit more tricky... we could use the proposal as above and have the engine-side stack API ask the client for different things...

or we could include all the auths that are needed for each service in the request... this could be the path of lease resistance, and on equally horrible footing as the existing solution.

Reply:
Also, this type is partly defined like this because it doesn't come from the body of a ServiceCreate request, but instead comes from the headers of the request.

Reply:
One option may be to have something like this in the `StackSpec`:

```go
type StackSpec struct {
    // ... elide the existing fields

    // RegistryAuths holds the information for registry auth tokens for various services, if needed.
    RegistryAuths RegsitryAuthSet `json:",omitempty"`
}

// RegistryAuthSet is a set of Registry credentials.
type RegistryAuthSet struct {
    // Credentials is a map of Service Name to Registry Auth Token needed for that service.
    Credentials map[string]string `json:",omitempty"`
}
```

Many other ways to implement it may be available. The information would, unfortunately, have to be sent as part of the request body, instead of a header.